### ISSUE TITLE
toJSON({ flat: true })

### ISSUE THREAD

**Issue by fabiancook** [virtualstate/focus#3]
## 🚀 Feature Proposal

Create a JSON object that represents the "most flat" the JSX tree can be represented as

## Motivation

To match a `fromJSON(tree, { flat: true })` implementation

## Example

```typescript jsx
const tree = (
  <main>
     <h1>Title</h1>
     <p>Paragraph 1</p>
     <p>Paragraph 2</p>
     <p>
        <h2>Paragraph 3</h2>
        Content
     </p>
     <footer>Footer</footer>
  </main>
)
```

```json
{
  "main": {
    "h1": "Title",
    "p": [
      "Paragraph 1",
      "Paragraph 2",
      [
        {
          "h2": "Paragraph 3"
        }, 
        "Content"
      ]
    ],
    "footer": "Footer"
  }
}
```

This may be a bad idea, and will reduce the information that we know about a JSX tree, for example the exact index of each node, but, it would be nice to be able to read a JSON object that looks like this, meaning, this module should be able to read it anyway, `fromJSON({flat: true})` should really just be `h(tree)`

**Comment by fabiancook** [comment]
This might be a nicer format

```JSON
{
  "main": [
     { "h1": "Title" },
     { "p": "Paragraph 1" },
     { "p": "Paragraph 2" },
     { "p": [
        { "h1": "Paragraph 3" },
        "Content"
     ]},
     { "footer": "Footer" }
  ]
}
```

**Comment by fabiancook** [comment]
My concern is how to deal with `properties`, but if a "standard" defined key is used specifically for them it will be fine

```json
[
  { "h1": "Title", "props": { "id": "title" }},
  { "p": "Paragraph 1", "attrs": { "id": "paragraph-1" }},
  { "p": "Paragraph 2", "attributes": { "id": "paragraph-2" }} 
]
```

```typescript
function isFlatTree(node: unknown) {
  if (!isUnknownJSXNode(node)) return false;
  const keys = Object.keys(node);
  if (!keys.length) return false;
  if (keys.length === 1) return true;
  const possible = keys.filter(key => !isPossibleJSXKey(key, possibleNameKeys));
  return possible.length === 1;
}

function isPossibleJSXKey(key: Key, excluding: Key[] = []) {
  return !!allPossibleKeys.find(possible => !excluding.includes(possible) && key === possible);
}
```

Here is a comparison between some string lengths

```typescript
JSON.stringify({
  "main": [
     { "h1": "Title" },
     { "p": "Paragraph 1" },
     { "p": "Paragraph 2" },
     { "p": [
        { "h1": "Paragraph 3" },
        "Content"
     ]},
     { "footer": "Footer" }
  ]
}).length
```

outputs `124`

```typescript
`<main><h1>Title</h1><p>Paragraph 1</p><p>Paragraph 2</p><p><h2>Paragraph 3</h2>\nContent</p><footer>Footer</footer></main>`.length
```

outputs `121`

```typescript
JSON.stringify({
  "name": "main",
   "chilren": [
     { "name": "h1", "children": "Title" },
     { "name": "p", "children": "Paragraph 1" },
     { "name": "p", "children": "Paragraph 2" },
     { "name": "p", "children": [
        { "h1": "Paragraph 3" },
        "Content"
     ]},
     { "name": "footer", "children": "Footer" }
  ]
}).length
```

outputs `231`

**Comment by fabiancook** [comment]
With [dde9bb9](https://github.com/virtualstate/focus/commit/dde9bb953b8fcea6ae9c8c007766e2820677ee90) `h` now supports the second format

```json
{
    "main": [
        { "h1": "Title", "props": { "id": "title" }},
        { "p": "Paragraph 1", "attrs": { "id": "paragraph-1" }},
        { "p": "Paragraph 2", "attributes": { "id": "paragraph-2" }},
        { "p": [
                { "h1": "Paragraph 3" },
                "Content"
            ]},
        { "footer": "Footer" }
    ]
}
```

Being the same as

```json
{
  "type": "main",
  "children": [
    {
      "type": "h1",
      "props": {
        "id": "title"
      },
      "children": [
        "Title"
      ]
    },
    {
      "type": "p",
      "props": {
        "id": "paragraph-1"
      },
      "children": [
        "Paragraph 1"
      ]
    },
    {
      "type": "p",
      "props": {
        "id": "paragraph-2"
      },
      "children": [
        "Paragraph 2"
      ]
    },
    {
      "type": "p",
      "children": [
        {
          "type": "h1",
          "children": [
            "Paragraph 3"
          ]
        },
        "Content"
      ]
    },
    {
      "type": "footer",
      "children": [
        "Footer"
      ]
    }
  ]
}
```

**Comment by fabiancook** [comment]
The second format also preserves all index information. The first proposed format would be confusing around how to define props!

**Comment by fabiancook** [comment]
Closed by [bd312e6](https://github.com/virtualstate/focus/commit/bd312e6155fdfb048e9f426f189603ddf47376a9)
